Sources/NIOFileSystem/FileSystem.swift

```diff
@@ -917,72 +923,228 @@ extension FileSystem {
+
+                        default:
+                            let error = FileSystemError(
+                                code: .unsupported,
+                                message: """
+                                    Can't copy '\(entry.path)' of type '\(entry.type)'; only regular \
+                                    files, symbolic links and directories can be copied.
+                                    """,
+                                cause: nil,
+                                location: .here()
+                            )
 
-                    case .symlink:
-                        if await shouldCopyFile(entrySource, entryDestination) {
-                            do {
-                                try await self.copySymbolicLink(
-                                    from: entry.path,
-                                    to: destinationPath.appending(entry.name)
-                                )
-                            } catch {
-                                try await shouldProceedAfterError(entry, error)
-                            }
+                            try await shouldProceedAfterError(entry, error)
                         }
+                    }
+                }
+            }
 
-                    case .directory:
-                        directoriesToCopy.append((entrySource, entryDestination))
+            contentsToCopy.append(.endOfDir)
+            return contentsToCopy
+        }
+    }
 
-                    default:
-                        let error = FileSystemError(
-                            code: .unsupported,
-                            message: """
-                                Can't copy '\(entrySource)' of type '\(entry.type)'; only regular \
-                                files, symbolic links and directories can be copied.
-                                """,
-                            cause: nil,
-                            location: .here()
+    /// This could be achieved through quite complicated special casing of the parallel copy.
+    /// The resulting code is far harder to read and debug though so this is kept as a special case
+    private func copyDirectorySequential(
+        from sourcePath: FilePath,
+        to destinationPath: FilePath,
+        shouldProceedAfterError: @escaping @Sendable (
+            _ entry: DirectoryEntry,
+            _ error: Error
+        ) async throws -> Void,
+        shouldCopyItem: @escaping @Sendable (
+            _ source: DirectoryEntry,
+            _ destination: FilePath
+        ) async -> Bool
+    ) async throws {
+        // Strategy: find all needed items to copy/recurse into while the directory is open;
+        // defer actual copying and recursion until after the source directory has been closed
+        // to avoid consuming too many file descriptors.
+        let toCopy = try await self.prepareDirectoryForRecusiveCopy(
```

Comment:
nit: can we call this `itemsToCopy` or something to avoid shadowing the `toCopy` enum case?